vault/eventbus/bus.go

```diff
@@ -0,0 +1,158 @@
+
+import (
+	"context"
+	"errors"
+	"net/url"
+	"strings"
+	"sync/atomic"
+
+	"github.com/hashicorp/eventlogger"
+	"github.com/hashicorp/eventlogger/formatter_filters/cloudevents"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/vault/sdk/logical"
+)
+
+var ErrNotStarted = errors.New("event broker has not been started")
+
+// EventBus contains the main logic of running an event broker for Vault.
+// Start() must be called before the EventBus will accept events for sending.
+type EventBus struct {
+	logger  hclog.Logger
+	broker  *eventlogger.Broker
+	started atomic.Bool
+}
+
+func NewEventBus() *EventBus {
+	return &EventBus{
+		logger: hclog.Default().Named("eventbus"),
+		broker: eventlogger.NewBroker(),
+	}
+}
+
+// Start starts the event bus, allowing events to be written.
+// It is not possible to stop or restart the event bus.
+// It is safe to call Start() multiple times.
+func (bus *EventBus) Start() {
+	bus.started.Store(true)
+}
+
+var _ logical.EventSender = (*EventBus)(nil)
+
+// Send sends an event to the event bus and routes it to all relevant subscribers.
+// This function does *not* wait for all subscribers to acknowledge before returning.
+// TODO: use schema once it is defined
+func (bus *EventBus) Send(ctx context.Context, eventType string, s any) error {
+	if !bus.started.Load() {
+		return ErrNotStarted
+	}
+	bus.logger.Info("Sending event", "event", s)
+	_, err := bus.broker.Send(ctx, eventlogger.EventType(eventType), s)
+	if err != nil {
+		// if no listeners for this event type are registered, that's okay, the event
+		// will just not be sent anywhere
+		if strings.Contains(strings.ToLower(err.Error()), "no graph for eventtype") {
+			return nil
+		}
+	}
+	return err
+}
+
+func (bus *EventBus) Subscribe(_ context.Context, eventType string) (chan any, error) {
+	// subscriptions are still stored even if the bus has not been started
+	pipelineID, err := uuid.GenerateUUID()
+	if err != nil {
+		return nil, err
+	}
+
+	nodeID, err := uuid.GenerateUUID()
+	if err != nil {
+		return nil, err
+	}
+	formatterNodeID, err := uuid.GenerateUUID()
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: should we have just one node, and handle all the routing ourselves?
+	asyncNode := newAsyncNode()
+	err = bus.broker.RegisterNode(eventlogger.NodeID(nodeID), asyncNode)
+	if err != nil {
+		defer asyncNode.Close()
+		return nil, err
+	}
+
+	err = bus.broker.RegisterNode(eventlogger.NodeID(formatterNodeID), cloudEventsFormatterFilter)
+	if err != nil {
+		defer asyncNode.Close()
+		return nil, err
+	}
+
+	nodes := []eventlogger.NodeID{eventlogger.NodeID(formatterNodeID), eventlogger.NodeID(nodeID)}
+
+	pipeline := eventlogger.Pipeline{
+		PipelineID: eventlogger.PipelineID(pipelineID),
+		EventType:  eventlogger.EventType(eventType),
+		NodeIDs:    nodes,
+	}
+	err = bus.broker.RegisterPipeline(pipeline)
+	if err != nil {
+		defer asyncNode.Close()
+		return nil, err
+	}
+	return asyncNode.ch, nil
```

Comment:
Thinking aloud a bit, I think we'll need to return more than just the channel when we implement unsubscribing, as we want the node to own closing the channel, so we need some other way of the subscriber sending a signal back to the event broker.